plugins/JsTrackerInstallCheck/JsTrackerInstallCheck.php

```diff
@@ -0,0 +1,149 @@
+        if (empty($nonceOptionString)) {
+            return;
+        }
+
+        $nonceOptionArray = json_decode($nonceOptionString, true);
+        if (empty($nonceOptionArray)) {
+            return;
+        }
+
+        // Make sure that the nonce matches
+        if (empty($nonceOptionArray['nonce']) || $nonceOptionArray['nonce'] !== $trackerInstallCheckParam) {
+            return;
+        }
+
+        // If the nonce is older 30 seconds, ignore it. This should be plenty of time because an API call creates the nonce just before opening the site
+        if (empty($nonceOptionArray['time']) || Date::getNowTimestamp() - $nonceOptionArray['time'] > 30) {
+            return;
+        }
+
+        // Since the nonce matches and hasn't expired, update the option indicating success
+        $nonceOptionArray['isSuccessful'] = true;
+        Option::set(self::OPTION_NAME_PREFIX . $request->getIdSite(), json_encode($nonceOptionArray));
+    }
+
+    /**
+     * Check whether a test request has been recorded for the provided nonce. If no nonce is provided, the recorded
+     * result for the site will be returned.
+     *
+     * @param string $idSite
+     * @param string $nonce The unique nonce used to identify the test requests. Optionally can be left empty if simply
+     * wanting to check if the site has been successfully tested.
+     * @return array list of containers ['isSuccess' => true]
+     */
+    public function checkForJsTrackerInstallTestSuccess(string $idSite, string $nonce = ''): array
+    {
+        $nonceOptionString = Option::get(self::OPTION_NAME_PREFIX . $idSite);
+        if (empty($nonceOptionString)) {
+            return ['isSuccess' => false];
+        }
+
+        $nonceOptionArray = json_decode($nonceOptionString, true);
+        if (empty($nonceOptionArray)) {
+            return ['isSuccess' => false];
+        }
+
+        // Check if the nonce matches the recorded nonce
+        if (!empty($nonce) && (empty($nonceOptionArray['nonce']) || $nonceOptionArray['nonce'] !== $nonce)) {
+            return ['isSuccess' => false];
+        }
+
+        return ['isSuccess' => !empty($nonceOptionArray['isSuccessful'])];
+    }
+
+    /**
+     * Initiate a test whether the JS tracking code has been successfully installed for a site. It generates a nonce and
+     * stores it in the option table so that it can be accessed later during the Tracker.isExcludedVisit event.
+     *
+     * @param string $idSite
+     * @return array containing the URL constructed using the main URL for the site and the newly created nonce as a
+     * query parameter.
+     * E.g ['url' => 'https://some-site.com?tracker_install_check=c3dfa1abbbab6381baca0793b8dd5d', 'nonce' => 'c3dfa1abbbab6381baca0793b8dd5d']
+     * @throws \Exception
+     */
+    public function initiateJsTrackerInstallTest(string $idSite): array
+    {
+        $nonceString = md5(SettingsPiwik::getSalt() . time() . Common::generateUniqId());
```

Comment:
So the nonce is updated each time this method is called, right? So if two different people trigger that around the same time for the same site, one of those might fail as the nonce was overwritten by the other one, right?

Reply:
Thank you @sgiehl . That's a very good point. I guess I could check the age of the existing nonce and if it's less than a few seconds old, simply reuse it.